crates/subspace-fraud-proof/src/invalid_transaction_proof.rs

```diff
@@ -0,0 +1,214 @@
+        Hash,
+        Exec,
+        VerifierClient,
+        DomainExtrinsicsBuilder,
+    >
+where
+    PBlock: BlockT,
+    Hash: Encode + Decode,
+    H256: Into<PBlock::Hash>,
+    PClient: HeaderBackend<PBlock> + ProvideRuntimeApi<PBlock> + Send + Sync,
+    PClient::Api: ExecutorApi<PBlock, Hash>,
+    VerifierClient: VerifierApi,
+    DomainExtrinsicsBuilder: BuildDomainExtrinsics<PBlock>,
+    Exec: CodeExecutor + 'static,
+{
+    /// Constructs a new instance of [`InvalidStateTransitionProofVerifier`].
+    pub fn new(
+        primary_chain_client: Arc<PClient>,
+        executor: Arc<Exec>,
+        verifier_client: VerifierClient,
+        domain_extrinsics_builder: DomainExtrinsicsBuilder,
+    ) -> Self {
+        Self {
+            primary_chain_client,
+            executor,
+            verifier_client,
+            domain_extrinsics_builder,
+            _phantom: Default::default(),
+        }
+    }
+
+    /// Verifies the invalid transaction proof.
+    pub fn verify(
+        &self,
+        invalid_transaction_proof: &InvalidTransactionProof,
+    ) -> Result<(), VerificationError> {
+        let InvalidTransactionProof {
+            domain_id,
+            block_number,
+            domain_block_hash,
+            invalid_extrinsic,
+            storage_proof,
+        } = invalid_transaction_proof;
+
+        // TODO: Validate bundle solution.
+        // - Bundle is valid and is produced by a legit executor.
+        // - Bundle author, who will be slashed, can be extracted in runtime.
+
+        let primary_hash: PBlock::Hash = self
+            .verifier_client
+            .primary_hash(*domain_id, *block_number)?
+            .into();
+
+        let header = self
+            .primary_chain_client
+            .header(primary_hash)?
+            .ok_or_else(|| {
+                sp_blockchain::Error::Backend(format!("Header for {primary_hash} not found"))
+            })?;
+        let primary_parent_hash = *header.parent_hash();
+
+        let domain_runtime_code = crate::domain_runtime_code::retrieve_domain_runtime_code(
+            *domain_id,
+            primary_parent_hash,
+            &self.primary_chain_client,
+        )?;
+
+        // TODO: Verifiable invalid extrinsic.
+        // Once the bundle validity is checked, we ought to know about `extrinsics_root`, plus a
+        // Merkle proof of invalid extrinsic, we can get the original invalid extrinsic data in a
+        // verifiable way.
+        let extrinsic = OpaqueExtrinsic::from_bytes(invalid_extrinsic)?;
+
+        // TODO: proper state_root
+        let state_root = H256::default();
+        let db = storage_proof.clone().into_memory_db::<BlakeTwo256>();
+        let read_value = |storage_key| {
+            read_trie_value::<LayoutV1<BlakeTwo256>, _>(&db, &state_root, storage_key, None, None)
+                .map_err(|_| VerificationError::InvalidStorageProof)
+        };
+
+        let next_fee_multiplier_storage_key = [
```

Comment:
This is not very pleasing to look at. We create a representation like storage. Here is for Receipts https://github.com/subspace/subspace/blob/main/domains/primitives/messenger/src/messages.rs#L284

Reply:
We would still need to change them manually in my proopsed one as well. Its easier to know what the structure is when the storage is defined. Raw bytes is always confusing IMO

Reply:
resolved TODO in this PR